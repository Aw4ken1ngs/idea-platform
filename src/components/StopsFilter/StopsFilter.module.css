.filterTitle {
    font-size: 16px;
    font-weight: bold;
    margin-bottom: 8px;
    color: #6c757d;
}

.checkboxGroup {
    display: flex;
    flex-direction: column;
    gap: 8px;
}

.checkboxLabel {
    display: flex;
    align-items: center;
    font-size: 14px;
    color: #6c757d;
    cursor: pointer;
}

.checkbox {
    margin-right: 8px;
    width: 16px;
    height: 16px;
    border: 1px solid #ccc;
    border-radius: 4px;
    appearance: none;
    background-color: #f9f9f9;
    cursor: pointer;
    transition: background-color 0.2s, border-color 0.2s;
}

.checkbox:checked {
    background-color: #007bff;
    border-color: #007bff;
    background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'%3E%3Cpolyline points='3,8 6,11 13,4' fill='none' stroke='%23fff' stroke-width='2'/%3E%3C/svg%3E");
    background-size: 12px 12px;
    background-repeat: no-repeat;
    background-position: center;
}
