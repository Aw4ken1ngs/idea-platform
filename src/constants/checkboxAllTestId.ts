export const CHECKBOX_ALL_TEST_ID:string = "checkbox-all"