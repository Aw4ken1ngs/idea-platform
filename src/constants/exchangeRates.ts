export const EXCHANGE_RATES = { RUB: 1, USD: 99.83, EUR: 103.98 };