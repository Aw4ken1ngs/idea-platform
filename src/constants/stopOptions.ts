export const STOP_OPTIONS = [0, 1, 2, 3];