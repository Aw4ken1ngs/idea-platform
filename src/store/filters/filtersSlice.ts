import {createSlice} from "@reduxjs/toolkit";
import type {PayloadAction} from "@reduxjs/toolkit";
import {Currency} from "../../types";

interface FiltersState {
    currency: Currency;
    stops: number[];
}

const initialState:FiltersState = {
    currency: Currency.RUB,
    stops: []
}

export const filtersSlice = createSlice({
    name: 'filters',
    initialState,
    reducers: {
        setCurrency: (state, action: PayloadAction<Currency>) => {
            state.currency = action.payload;
        },
        setStops: (state, action: PayloadAction<number[]>) => {
            state.stops = action.payload;
        }
    }
})

export const { setCurrency, setStops } = filtersSlice.actions
export default filtersSlice.reducer