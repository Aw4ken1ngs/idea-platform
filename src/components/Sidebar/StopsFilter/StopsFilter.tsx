import React from "react";
import styles from "./StopsFilter.module.css";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../../store/store";
import { setStops } from "../../../store/filters/filtersSlice";

const StopsFilter: React.FC = () => {
    const selectedStops = useSelector((store: RootState) => store.filters.stops);
    const dispatch = useDispatch();

    const handleCheckboxChange = (stop: number, checked: boolean) => {
        if (checked) {
            dispatch(setStops([...selectedStops, stop]));
        } else {
            dispatch(setStops(selectedStops.filter((s) => s !== stop)));
        }
    };

    return (
        <div className={styles.stopsFilter}>
            <h3 className={styles.filterTitle}>Количество пересадок:</h3>
            <div className={styles.checkboxGroup}>
                <label className={styles.checkboxLabel}>
                    <input
                        type="checkbox"
                        checked={selectedStops.length === 0}
                        onChange={() => dispatch(setStops([]))}
                        className={styles.checkbox}
                        data-testid="checkbox-all"
                    />
                    Все
                </label>
                {[0, 1, 2, 3].map((stop) => (
                    <label key={stop} className={styles.checkboxLabel}>
                        <input
                            type="checkbox"
                            checked={selectedStops.includes(stop)}
                            onChange={(e) => handleCheckboxChange(stop, e.target.checked)}
                            className={styles.checkbox}
                            data-testid={`checkbox-${stop}`}
                        />
                        {stop === 0 ? "Без пересадок" : `${stop} пересадка(и)`}
                    </label>
                ))}
            </div>
        </div>
    );
};

export default StopsFilter;
