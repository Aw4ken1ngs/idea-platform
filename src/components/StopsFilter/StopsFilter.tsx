import React from "react";
import styles from "./StopsFilter.module.css";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../store/store";
import { setStops } from "../../store/filters/filtersSlice";

const StopsFilter: React.FC = () => {
    const selectedStops = useSelector((store: RootState) => store.filters.stops);
    const dispatch = useDispatch();

    const handleCheckboxChange = (stop: number, checked: boolean) => {
        if (checked) {
            dispatch(setStops([...selectedStops, stop]));
        } else {
            dispatch(setStops(selectedStops.filter((s) => s !== stop)));
        }
    };

    return (
        <div className={styles.stopsFilter}>
            <h3 className={styles.filterTitle}>Количество пересадок:</h3>
            <div className={styles.checkboxGroup}>
                <label className={styles.checkboxLabel}>
                    <input
                        type="checkbox"
                        checked={selectedStops.length === 0}
                        onChange={() => dispatch(setStops([]))}
                        className={styles.checkbox}
                    />
                    Все
                </label>
                {[0, 1, 2, 3].map((stop) => (
                    <label key={stop} className={styles.checkboxLabel}>
                        <input
                            type="checkbox"
                            checked={selectedStops.includes(stop)}
                            onChange={(e) => handleCheckboxChange(stop, e.target.checked)}
                            className={styles.checkbox}
                        />
                        {stop === 0 ? "Без пересадок" : `${stop} пересадка(и)`}
                    </label>
                ))}
            </div>
        </div>
    );
};

export default StopsFilter;



// import React from "react";
// import styles from "./StopsFilter.module.css";
// import {useDispatch, useSelector} from "react-redux";
// import {RootState} from "../../store/store";
// import {setStops} from "../../store/filters/filtersSlice";
//
// const StopsFilter: React.FC= () => {
//     const selectedStops = useSelector((store:RootState) => {
//         return store.filters.stops
//     } )
//     const dispatch = useDispatch();
//     const handleCheckboxChange = (stop: number, checked: boolean) => {
//         if (checked) {
//             dispatch(setStops([...selectedStops, stop]));
//         } else {
//             dispatch(setStops(selectedStops.filter((s) => s !== stop)));
//         }
//     };
//
//     return (
//         <div className={styles.stopsFilter}>
//             <h3>Фильтр по пересадкам:</h3>
//             <label>
//                 <input
//                     type="checkbox"
//                     checked={selectedStops.length === 0}
//                     onChange={() => dispatch(setStops([]))}
//                 />
//                 Все
//             </label>
//             {[0, 1, 2, 3].map((stop) => (
//                 <label key={stop}>
//                     <input
//                         type="checkbox"
//                         checked={selectedStops.includes(stop)}
//                         onChange={(e) => handleCheckboxChange(stop, e.target.checked)}
//                     />
//                     {stop === 0 ? "Без пересадок" : `${stop} пересадка(и)`}
//                 </label>
//             ))}
//         </div>
//     );
// };
//
// export default StopsFilter;
