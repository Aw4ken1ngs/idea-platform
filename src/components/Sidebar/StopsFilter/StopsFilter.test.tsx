import { configureStore } from "@reduxjs/toolkit";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Provider } from "react-redux";
import * as filtersSlice from "../../../store/filters/filtersSlice";
import { Currency } from "../../../types";
import StopsFilter from "./StopsFilter";
import {CHECKBOX_ALL_TEST_ID} from "../../../constants/checkboxAllTestId";

const store = configureStore({
    reducer: {
        filters: filtersSlice.default,
    },
    preloadedState: {
        filters: {
            currency: Currency.RUB,
            stops: [],
        },
    },
});

const renderComponent = () => {
    return (
        <Provider store={store}>
            <StopsFilter />
        </Provider>
    );
};

describe("<StopsFilter />", () => {
    const user = userEvent.setup();

    it("should match snapshot", () => {
        const { container } = render(renderComponent());
        expect(container).toMatchSnapshot();
    });

    test('should check that the checkbox "All" is changing', async () => {
        const { getByTestId } = render(renderComponent());

        const allCheckbox = getByTestId(CHECKBOX_ALL_TEST_ID) as HTMLInputElement;


        expect(allCheckbox.checked).toBe(true);

        const noStopsCheckbox = getByTestId("checkbox-0") as HTMLInputElement;

        await user.click(noStopsCheckbox);

        expect(allCheckbox.checked).toBe(false);
        expect(noStopsCheckbox.checked).toBe(true);

    });
    test('should allow selecting "No stops" and "1 stop" only', async () => {
        const { getByTestId } = render(renderComponent());


        const noStopsCheckbox = getByTestId("checkbox-0") as HTMLInputElement;
        const oneStopCheckbox = getByTestId("checkbox-1") as HTMLInputElement;
        const allCheckbox = getByTestId(CHECKBOX_ALL_TEST_ID) as HTMLInputElement;
        const twoStopsCheckbox = getByTestId("checkbox-2") as HTMLInputElement;
        const threeStopsCheckbox = getByTestId("checkbox-3") as HTMLInputElement;

        expect(noStopsCheckbox.checked).toBe(true);


        await user.click(oneStopCheckbox);
        expect(oneStopCheckbox.checked).toBe(true);

        expect(allCheckbox.checked).toBe(false);
        expect(twoStopsCheckbox.checked).toBe(false);
        expect(threeStopsCheckbox.checked).toBe(false);
    });

    test('should uncheck all other checkboxes when "All" is selected', async () => {
        const { getByTestId } = render(renderComponent());

        const noStopsCheckbox = getByTestId("checkbox-0") as HTMLInputElement;
        const oneStopCheckbox = getByTestId("checkbox-1") as HTMLInputElement;
        const allCheckbox = getByTestId(CHECKBOX_ALL_TEST_ID) as HTMLInputElement;
        const twoStopsCheckbox = getByTestId("checkbox-2") as HTMLInputElement;
        const threeStopsCheckbox = getByTestId("checkbox-3") as HTMLInputElement;

        await user.click(twoStopsCheckbox);


        expect(noStopsCheckbox.checked).toBe(true);
        expect(oneStopCheckbox.checked).toBe(true);
        expect(twoStopsCheckbox.checked).toBe(true);
        expect(allCheckbox.checked).toBe(false);


        await user.click(allCheckbox);

        expect(allCheckbox.checked).toBe(true);
        expect(noStopsCheckbox.checked).toBe(false);
        expect(oneStopCheckbox.checked).toBe(false);
        expect(twoStopsCheckbox.checked).toBe(false);
        expect(threeStopsCheckbox.checked).toBe(false);
    });

    test("should dispatch an action and update state when a checkbox is changed", async () => {
        const { getByTestId } = render(renderComponent());

        const setStopsSpy = jest.spyOn(filtersSlice, "setStops");

        const noStopsCheckbox = getByTestId("checkbox-0") as HTMLInputElement;
        const oneStopCheckbox = getByTestId("checkbox-1") as HTMLInputElement;

        await user.click(noStopsCheckbox);

        expect(setStopsSpy).toBeCalledTimes(1);
        expect(setStopsSpy).toHaveBeenCalledWith([0]);

        expect(noStopsCheckbox.checked).toBe(true);

        await user.click(oneStopCheckbox);

        expect(setStopsSpy).toBeCalledTimes(2);
        expect(setStopsSpy).toHaveBeenCalledWith([0, 1]);

        expect(noStopsCheckbox.checked).toBe(true);
        expect(oneStopCheckbox.checked).toBe(true);

        await user.click(noStopsCheckbox);

        expect(setStopsSpy).toBeCalledTimes(3);
        expect(setStopsSpy).toHaveBeenCalledWith([1]);

        expect(noStopsCheckbox.checked).toBe(false);
    });
});

