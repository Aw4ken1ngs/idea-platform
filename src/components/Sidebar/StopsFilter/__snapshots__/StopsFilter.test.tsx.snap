// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StopsFilter /> should match snapshot 1`] = `
<div>
  <div
    class="stopsFilter"
  >
    <h3
      class="filterTitle"
    >
      Количество пересадок:
    </h3>
    <div
      class="checkboxGroup"
    >
      <label
        class="checkboxLabel"
      >
        <input
          checked=""
          class="checkbox"
          data-testid="checkbox-all"
          type="checkbox"
        />
        Все
      </label>
      <label
        class="checkboxLabel"
      >
        <input
          class="checkbox"
          data-testid="checkbox-0"
          type="checkbox"
        />
        Без пересадок
      </label>
      <label
        class="checkboxLabel"
      >
        <input
          class="checkbox"
          data-testid="checkbox-1"
          type="checkbox"
        />
        1 пересадка(и)
      </label>
      <label
        class="checkboxLabel"
      >
        <input
          class="checkbox"
          data-testid="checkbox-2"
          type="checkbox"
        />
        2 пересадка(и)
      </label>
      <label
        class="checkboxLabel"
      >
        <input
          class="checkbox"
          data-testid="checkbox-3"
          type="checkbox"
        />
        3 пересадка(и)
      </label>
    </div>
  </div>
</div>
`;
